---
title: "README"
author: "Kimberle Shen"
date: "9/28/2020"
output:
  md_document:
    variant: markdown_github
---

# rfca

This package contains functions that trains a Random Forest Model with a labelled Seurat Object, for predicting cell types/states in unlabelled datasets. It also contains a pre-trained Random Forest model, as well as example datasets. 
# Introduction

Manual cell annotation of scRNAseq datasets, typically based on marker genes, can be time-consuming and biased. Being able to automatically predict cell types/states in a cell-by-cell and cluster-unbiased way is useful for fast and accurate phenotyping. 

In addition, despite the increasing amounts of scRNAseq datasets being generated, thorough analyis of these datasets is lagging, and/or done in silos. This package comes with a preloaded Random Forest model based on different datasets and cell types/states, that will be constantly updated. 

# Installation

You can install this package from GitHub with `devtools::install_github("kimberle9/rfca")`

# Examples

## Example 1: Creating a Seurat Object

```{r eval = FALSE}
# Load the rfca and Seurat libraries, as well as example datasets
library(rfca)
library(Seurat)
data("exampleSeuratObjectUnlabelled")
data("exampleSeuratObjectLabelled")

# Create Seurat Object with PCA and UMAP calculated
mySeuratObject <- createSeuratObjectPipeline(data.dir = "~/filtered_feature_bc_matrix", 
                                             nFeature_RNA_lower = 500, 
                                             nFeature_RNA_upper = 5000, 
                                             percent.mt = 5, 
                                             nfeatures = 2000, 
                                             dims = 20, 
                                             clusterResolution = 0.8)

# Assign cell type/state Idents to mySeuratObject manually, if you want to use it as a training dataset
```

## Example 2: Using Random Forest to train and predict cell types

```{r}
library(rfca)
library(Seurat)
data("exampleSeuratObjectUnlabelled")
data("exampleSeuratObjectLabelled")

# Create Random Forest Model with your labelled Seurat Object
myRandomForestModel <- createRFModel(exampleSeuratObjectLabelled)

# Create marker gene list from random forest model
markerGeneList <- createGeneLists(myRandomForestModel)

# Visualize Feature Plot based on marker gene list
myPlot <- cellMarkerPlots(exampleSeuratObjectLabelled, geneList = markerGeneList)
myPlot

# Predict cells based on your own created Random Forest Model created above
autoLabelledSeuratObject <- predictCells(exampleSeuratObjectUnlabelled, myRandomForestModel)

```

