<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to rfca â€¢ rfca</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to rfca">
<meta property="og:description" content="This vignette documents the functions in the rfca package. 
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rfca</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette.html">Introduction to rfca</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to rfca</h1>
                        <h4 class="author">Kimberle Shen</h4>
            
            <h4 class="date">9/30/2020</h4>
      
      
      <div class="hidden name"><code>vignette.Rmd</code></div>

    </div>

    
    
<div id="rfca" class="section level1">
<h1 class="hasAnchor">
<a href="#rfca" class="anchor"></a>rfca</h1>
<p>The rfca package contains functions that trains a Random Forest Model from single-cell RNA-Seq data, and allows the prediction of cell types and cell states in unlabelled datasets. It also contains a pre-trained Random Forest model, as well as example datasets.</p>
<p>This can be used for automatically labelling cell types for downstream applications and subsetting, looking at proportions of cell subtypes, and finding similarities in cell types, sub cell types, or cell states in different datasets.</p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Manual cell annotation of scRNAseq datasets, typically based on marker genes, can be time-consuming and biased. Being able to automatically predict cell types/states in a cell-by-cell and cluster-unbiased way is useful for fast and accurate phenotyping.</p>
<p>In addition, despite the increasing amounts of scRNAseq datasets being generated, thorough analysis of these datasets is lagging, and/or done in silos. This package comes with a preloaded Random Forest model based on different datasets and cell types/states, that will be constantly updated.</p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>You can install this package from GitHub with <code><a href="https://devtools.r-lib.org//reference/remote-reexports.html">devtools::install_github("kimberleshen/rfca")</a></code></p>
</div>
<div id="functions" class="section level1">
<h1 class="hasAnchor">
<a href="#functions" class="anchor"></a>Functions</h1>
<div id="load-packages-and-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#load-packages-and-examples" class="anchor"></a>Load packages and examples</h2>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rfca</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"exampleSeuratObjectUnlabelled"</span><span class="op">)</span> <span class="co"># example dataset</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"exampleSeuratObjectLabelled"</span><span class="op">)</span> <span class="co"># example dataset</span></pre></div>
</div>
<div id="createseuratobjectpipeline" class="section level2">
<h2 class="hasAnchor">
<a href="#createseuratobjectpipeline" class="anchor"></a>createSeuratObjectPipeline()</h2>
<p><code><a href="../reference/createSeuratObjectPipeline.html">createSeuratObjectPipeline()</a></code> is a function that creates a <code>Seurat</code> object from 10X cell ranger output, specifically the filtered_feature_bc_matrix directory containing the three files: barcodes.tsv.gz, matrix.mtx.gz, features.tsv.gz. It performs several steps in the <code>Seurat</code> pipeline, and returns a <code>Seurat</code> object with a computer pca and umap clusters as Idents (identities). These cluster Idents may be used as labels for training and building the Random Forest model with <code><a href="../reference/createRFModel.html">createRFModel(seuratObject)</a></code> to identify similar cells in other datasets. Alternatively, Idents (such as cell type) can be manually added and used as labels.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># Create Seurat Object with PCA and UMAP calculated</span>
<span class="va">mySeuratObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createSeuratObjectPipeline.html">createSeuratObjectPipeline</a></span><span class="op">(</span>data.dir <span class="op">=</span> <span class="st">"~/filtered_feature_bc_matrix"</span>, 
                                             nFeature_RNA_lower <span class="op">=</span> <span class="fl">500</span>, 
                                             nFeature_RNA_upper <span class="op">=</span> <span class="fl">5000</span>, 
                                             percent.mt <span class="op">=</span> <span class="fl">5</span>, 
                                             nfeatures <span class="op">=</span> <span class="fl">2000</span>, 
                                             dims <span class="op">=</span> <span class="fl">20</span>, 
                                             clusterResolution <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span></pre></div>
</div>
<div id="createrfmodel" class="section level2">
<h2 class="hasAnchor">
<a href="#createrfmodel" class="anchor"></a>createRFModel()</h2>
<p>This function should be used when there is a well-labelled dataset containing cell types you want to predict. <code><a href="../reference/createRFModel.html">createRFModel()</a></code> is a function that generates a Random Forest model from a labelled Seurat Object, with each cell as a datapoint, and gene expression levels of the most variable genes (HVGs) as features. The default <code>nfeatures</code> is 200. The Idents (use <code><a href="https://rdrr.io/r/base/table.html">table(Idents(seuratObject))</a></code>) of each cell are its labels, and the values that will be predicted when the model is used in <code><a href="../reference/predictCells.html">predictCells(seuratObject, theRFModel)</a></code>. A low OOB error rate of the Random Forest model (&lt;5%) implies the classified cell types from the input training dataset are well defined and separated.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="co"># Create Random Forest Model with your labelled Seurat Object</span>
<span class="va">myRandomForestModel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createRFModel.html">createRFModel</a></span><span class="op">(</span><span class="va">exampleSeuratObjectLabelled</span>, nfeatures <span class="op">=</span> <span class="fl">200</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Call:</span>
<span class="co">#&gt;  randomForest(formula = V1 ~ ., data = merged, importance = TRUE,      proximity = TRUE) </span>
<span class="co">#&gt;                Type of random forest: classification</span>
<span class="co">#&gt;                      Number of trees: 500</span>
<span class="co">#&gt; No. of variables tried at each split: 13</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;         OOB estimate of  error rate: 4.21%</span>
<span class="co">#&gt; Confusion matrix:</span>
<span class="co">#&gt;                  Astrocytes Endothelial Microglia Neurons Oligodendrocytes OPCs</span>
<span class="co">#&gt; Astrocytes               46           0         0       0                1    0</span>
<span class="co">#&gt; Endothelial               0          54         0       0                0    0</span>
<span class="co">#&gt; Microglia                 0           0        47       0                0    0</span>
<span class="co">#&gt; Neurons                   1           0         0      46                1    3</span>
<span class="co">#&gt; Oligodendrocytes          0           0         0       1               46    0</span>
<span class="co">#&gt; OPCs                      0           0         1       2                1   47</span>
<span class="co">#&gt; Tcells                    0           1         1       1                0    0</span>
<span class="co">#&gt; VSMCs                     1           0         1       0                0    0</span>
<span class="co">#&gt;                  Tcells VSMCs class.error</span>
<span class="co">#&gt; Astrocytes            0     0  0.02127660</span>
<span class="co">#&gt; Endothelial           0     0  0.00000000</span>
<span class="co">#&gt; Microglia             0     0  0.00000000</span>
<span class="co">#&gt; Neurons               1     0  0.11538462</span>
<span class="co">#&gt; Oligodendrocytes      0     0  0.02127660</span>
<span class="co">#&gt; OPCs                  0     0  0.07843137</span>
<span class="co">#&gt; Tcells               49     0  0.05769231</span>
<span class="co">#&gt; VSMCs                 0    52  0.03703704</span></pre></div>
</div>
<div id="markergenelist" class="section level2">
<h2 class="hasAnchor">
<a href="#markergenelist" class="anchor"></a>markerGeneList()</h2>
<p>Given a Random Forest model, the <code>markerGeneList()</code> function is used to create lists of marker genes for each of the labels in the model. The dafault number of genes to return is 10. For each label in the Random Forest model, the top 10 most important genes for that label is returned, with the label as the name of that gene list. The returned list of list can be used in the <code><a href="../reference/cellMarkerPlots.html">cellMarkerPlots()</a></code> function to visualize the combined/average expression of a set of genes in a single UMAP plot.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co"># Create marker gene list from random forest model</span>
<span class="va">markerGeneList</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/createGeneLists.html">createGeneLists</a></span><span class="op">(</span><span class="va">myRandomForestModel</span><span class="op">)</span>
<span class="co">#&gt; [1] "Adding gene list for Astrocytes"</span>
<span class="co">#&gt; [1] "Adding gene list for Endothelial"</span>
<span class="co">#&gt; [1] "Adding gene list for Microglia"</span>
<span class="co">#&gt; [1] "Adding gene list for Neurons"</span>
<span class="co">#&gt; [1] "Adding gene list for Oligodendrocytes"</span>
<span class="co">#&gt; [1] "Adding gene list for OPCs"</span>
<span class="co">#&gt; [1] "Adding gene list for Tcells"</span>
<span class="co">#&gt; [1] "Adding gene list for VSMCs"</span></pre></div>
</div>
<div id="cellmarkerplots" class="section level2">
<h2 class="hasAnchor">
<a href="#cellmarkerplots" class="anchor"></a>cellMarkerPlots()</h2>
<p>The <code><a href="../reference/cellMarkerPlots.html">cellMarkerPlots()</a></code> function takes in a <code>Seurat</code> object and a list of gene list (list of list) and calculates the average expression of each gene list for each cell. A feature plot will be generated for each gene list, titled with the name of the name of the gene list, usually the labels of the Random Forest model if the list of gene list is generated with <code>createGeneList()</code>. Alternatively, the list of gene lists can be generated manually (eg. with your favorite cell type markers) and can be used for quicker annotation of clusters in an unlabelled dataset.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co"># Visualize Feature Plot based on marker gene list</span>
<span class="va">myPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cellMarkerPlots.html">cellMarkerPlots</a></span><span class="op">(</span><span class="va">exampleSeuratObjectLabelled</span>, geneList <span class="op">=</span> <span class="va">markerGeneList</span><span class="op">)</span>
<span class="co">#&gt; [1] "Calculating average expression of genes in Astrocytes list."</span>
<span class="co">#&gt; [1] "Calculating average expression of genes in Endothelial list."</span>
<span class="co">#&gt; [1] "Calculating average expression of genes in Microglia list."</span>
<span class="co">#&gt; [1] "Calculating average expression of genes in Neurons list."</span>
<span class="co">#&gt; [1] "Calculating average expression of genes in Oligodendrocytes list."</span>
<span class="co">#&gt; [1] "Calculating average expression of genes in OPCs list."</span>
<span class="co">#&gt; [1] "Calculating average expression of genes in Tcells list."</span>
<span class="co">#&gt; [1] "Calculating average expression of genes in VSMCs list."</span>
<span class="va">myPlot</span></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co"># Make your own marker gene list</span>
<span class="va">myList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">myList</span><span class="op">$</span><span class="va">Oligodendrocytes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Mbp"</span>, <span class="st">"Mobp"</span>, <span class="st">"Plp1"</span><span class="op">)</span>
<span class="va">myList</span><span class="op">$</span><span class="va">Microglia</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Iba1"</span>, <span class="st">"Cst7"</span><span class="op">)</span></pre></div>
</div>
<div id="predictcells" class="section level2">
<h2 class="hasAnchor">
<a href="#predictcells" class="anchor"></a>predictCells()</h2>
<p>This function takes a <code>Seurat</code> object and predicts the cell type for each cell is in a cluster-unbiased way. It returns the <code>Seurat</code> object with new labels as Idents. The Random Forest model used can be:</p>
<ol style="list-style-type: decimal">
<li>User-specified, where labels can be anything: cell types, cell subtypes, cell states, cluster identities etc.</li>
<li>Default models (currently only for mouse brain tissues). In the future, there will be several default models for different tissue types (mouse brain, mouse immune, mouse lung, human brain etc) to choose from (eg. <code>predictCells(exampleSeuratObjectUnlabelled, tissueType = mouseImmune))</code>
</li>
</ol>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co"># Predict cells based on your own Random Forest Model created above</span>
<span class="va">autoLabelledSeuratObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predictCells.html">predictCells</a></span><span class="op">(</span><span class="va">exampleSeuratObjectUnlabelled</span>, <span class="va">myRandomForestModel</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;       Astrocytes      Endothelial        Microglia          Neurons </span>
<span class="co">#&gt;               55               45               58               44 </span>
<span class="co">#&gt; Oligodendrocytes             OPCs           Tcells            VSMCs </span>
<span class="co">#&gt;               55               49               48               42</span>

<span class="co"># Predict cells based on my pre-loaded and pre-trained Random Forest Model</span>
<span class="va">autoLabelledSeuratObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predictCells.html">predictCells</a></span><span class="op">(</span><span class="va">exampleSeuratObjectUnlabelled</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;       Astrocytes      Endothelial        Microglia          Neurons </span>
<span class="co">#&gt;               53               46               53               48 </span>
<span class="co">#&gt; Oligodendrocytes             OPCs           Tcells            VSMCs </span>
<span class="co">#&gt;               53               49               48               46</span>

<span class="co"># Visualize autoLabelledSeuratObject</span>
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">autoLabelledSeuratObject</span><span class="op">)</span></pre></div>
<p><img src="vignette_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div id="datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#datasets" class="anchor"></a>Datasets</h1>
<p>The example datasets included, <code>exampleSeuratObjectUnlabelled</code> and <code>exampleSeuratObjectLabelled</code> are from mouse brain tissues. They are ~400 cells each and are in the <code>Seurat</code> object format. For <code>exampleSeuratObjectUnlabelled</code>, Idents are cluster numbers. For <code>exampleSeuratObjectLabelled</code>, Idents are cell types.</p>
</div>
<div id="examples" class="section level1">
<h1 class="hasAnchor">
<a href="#examples" class="anchor"></a>Examples</h1>
<p>Some examples can be found in the README.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kimberle Shen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
